builtin_array.go

```diff
@@ -1224,6 +1224,188 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
+			} else {
+				fromValue = src.values[pk]
+			}
+			a = append(a, fromValue)
+		}
+		return r.newArrayValues(a)
+	} else {
+		a := r.newArrayLength(length)
+		for k := int64(0); k < length; k++ {
+			pk := valueInt(k)
+			var fromValue Value
+			if k == actualIndex {
+				fromValue = value
+			} else {
+				fromValue = o.self.getIdx(pk, nil)
+			}
+			createDataPropertyOrThrow(a, pk, fromValue)
+		}
+		return a
+	}
+}
+
+func (r *Runtime) arrayproto_toReversed(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+
+	if src := r.checkStdArrayObj(o); src != nil {
+		a := make([]Value, 0, length)
+		for k := int64(0); k < length; k++ {
+			from := valueInt(length - k - 1)
+			fromValue := src.values[from]
+			a = append(a, fromValue)
+		}
+		return r.newArrayValues(a)
+	} else {
+		a := r.newArrayLength(length)
+		for k := int64(0); k < length; k++ {
+			pk := valueInt(k)
+			from := valueInt(length - k - 1)
+			fromValue := o.self.getIdx(from, nil)
+			createDataPropertyOrThrow(a, pk, fromValue)
+		}
+		return a
+	}
+}
+
+func (r *Runtime) arrayproto_toSorted(call FunctionCall) Value {
+	var compareFn func(FunctionCall) Value
+	arg := call.Argument(0)
+	if arg != _undefined {
+		if arg, ok := arg.(*Object); ok {
+			compareFn, _ = arg.self.assertCallable()
+		}
+		if compareFn == nil {
+			panic(r.NewTypeError("The comparison function must be either a function or undefined"))
+		}
+	}
+
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+	if length >= math.MaxUint32 {
+		panic(r.newError(r.getRangeError(), "Invalid array length"))
+	}
+	var a []Value
+
+	if src := r.checkStdArrayObj(o); src != nil {
+		a = make([]Value, length)
+		copy(a, src.values)
+	} else {
+		a = make([]Value, 0, length)
+		for i := int64(0); i < length; i++ {
+			idx := valueInt(i)
+			a = append(a, nilSafe(o.self.getIdx(idx, nil)))
+		}
+	}
+
+	ar := r.newArrayValues(a)
+	ctx := arraySortCtx{
+		obj:     ar.self,
+		compare: compareFn,
+	}
+
+	sort.Stable(&ctx)
+	return ar
+}
+
+func (r *Runtime) arrayproto_toSpliced(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+	actualStart := relToIdx(call.Argument(0).ToInteger(), length)
+	var actualSkipCount int64
+	if len(call.Arguments) == 1 {
+		actualSkipCount = length - actualStart
+	} else if len(call.Arguments) > 1 {
+		actualSkipCount = min(max(call.Argument(1).ToInteger(), 0), length-actualStart)
+	}
+	itemCount := max(int64(len(call.Arguments)-2), 0)
+	newLength := length - actualSkipCount + itemCount
+	if newLength >= maxInt {
+		panic(r.NewTypeError("Invalid array length"))
+	}
+
+	if src := r.checkStdArrayObj(o); src != nil {
+		var values []Value
+		if itemCount < actualSkipCount {
+			values = make([]Value, len(src.values))
+			copy(values, src.values)
+			copy(values[actualStart+itemCount:], values[actualStart+actualSkipCount:])
+			values = values[:newLength]
```

Comment:
```suggestion
			values = make([]Value, newLength)
			copy(values, src.values[actualStart])
			copy(values[actualStart+itemCount:], src.values[actualStart+actualSkipCount:])
```

Haven't checked the code - but there is no need to overt allocate to then cut down.